supernode/p2p/service.go

```diff
@@ -0,0 +1,70 @@
+package p2p
+
+import (
+	"context"
+	"os"
+	"os/signal"
+
+	"github.com/pastelnetwork/gonode/common/log"
+)
+
+const (
+	logPrefix = "p2p"
+)
+
+// Service represents the p2p service.
+type Service struct {
+	dht DHT
+}
+
+// DHT represents the methods by which a library consumer interacts with a DHT.
+type DHT interface {
+	Store(data []byte, key []byte) (id string, err error)
+	Get(key string) (data []byte, found bool, err error)
+	CreateSocket() error
+	Listen() error
+	GetNetworkAddr() string
+	Bootstrap() error
+	UseStun() bool
+	Disconnect() error
+}
+
+// Run starts the DHT service
+func (service *Service) Run(ctx context.Context) error {
+	ctx = log.ContextWithPrefix(ctx, logPrefix)
+
+	err := service.dht.CreateSocket()
+	if err != nil {
+		panic(err)
+	}
+
+	go func() {
+		log.WithContext(ctx).Infof("Server listening on %q", service.dht.GetNetworkAddr())
+		err := service.dht.Listen()
+		panic(err)
+	}()
+
+	if err := service.dht.Bootstrap(); err != nil {
+		return err
+	}
+
+	c := make(chan os.Signal, 1)
+	signal.Notify(c, os.Interrupt)
```

Comment:
we catch signal in `cmd/app.go` [here](https://github.com/pastelnetwork/gonode/blob/2d9b6238f77f6dd2389bfd6aa47e7076cc760460/supernode/cmd/app.go#L104), so you can't use it. use `context` instead.
see [example](https://github.com/pastelnetwork/gonode/blob/2d9b6238f77f6dd2389bfd6aa47e7076cc760460/walletnode/api/server.go#L70-L100)